src/MiniCover/HitServices/HitServiceWithTests.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace MiniCover
+{
+    public static class HitServiceWithTests
+    {
+        private static readonly object lockObject = new object();
+        private static Dictionary<string, Dictionary<int, HitTrace>> files = new Dictionary<string, Dictionary<int, HitTrace>>();
+        private static Dictionary<string, bool> assemblyPdb = new Dictionary<string, bool>();
+        public static void Init(string fileName)
+        {
+            lock (lockObject)
+            {
+                if (!files.ContainsKey(fileName))
+                {
+                    files[fileName] = new Dictionary<int, HitTrace>();
+                }
+            }
+        }
+
+        public static void Hit(string fileName, int id)
+        {
+            lock (lockObject)
+            {
+                StackTrace stackTrace = new StackTrace();
+
+                var testMethod = GetTestMethod(stackTrace);
+                
+                var hits = files[fileName];
+                if (!hits.ContainsKey(id))
+                    hits.Add(id, new HitTrace(id));
+                hits[id].Hited(testMethod);
+            }
+        }
+
+        private static MethodBase GetTestMethod(StackTrace stack)
+        {
+            var frames = stack.GetFrames();
+            for (int i = frames.Length - 1; i >= 0; i--)
+            {
+                var currentMethod = frames[i].GetMethod();
+                if (currentMethod.HasPdb())
+                    return currentMethod;
+            }
+
+            return null;
+        }
+
+        private static bool HasPdb(this MethodBase methodBase)
+        {
+            var location = methodBase.DeclaringType.Assembly.Location;
+
+            if (!assemblyPdb.TryGetValue(location, out var hasPdb))
+            {
+                assemblyPdb[location] = hasPdb = File.Exists(Path.ChangeExtension(location, ".pdb"));
```

Comment:
Good idea! I was checking the properties on StackFrame, and found some that might have the same effect:
`.GetFileName()`, `.GetFileLineNumber()`
Checking if GetFileName doesn't return null might make your code simpler.
PS: I don't know if that method throws exception, or have bad performance

Reply:
GetFileName() give nothing that's why I used this